Channels 2 support

facucosta opened the pull request:
These changes give beatserver Channels 2 support without dropping Channels 1 support. Admittedly the modifications are a bit hacky due to my inexperience and the difference between Channels 1 and Channels 2 APIs, so maybe it would make more sense to fork the project into a Channels-1 version and a Channels-2 version to keep the code clean...

In addition to the changes in these files, the beatserver.py file requires additional lines to work correctly in Channels 2 (no changes needed if the user wants to keep working with Channels 1) which should be documented in the README file. This is an example beatserver.py file that works with Channels 2:

import os
from datetime import timedelta
from channels.layers import get_channel_layer

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "PROJECT_NAME.settings")

BEAT_SCHEDULE = {
    'send-every-5-seconds': {
        'channel_name': 'CHANNEL_NAME',
        'schedule': timedelta(seconds=5),
        'message': {'type': 'type_of_message', 'message': 'message_text'}
    },
}

channel_layers = get_channel_layer()

and the server needs to be initialized with the following command:

beatserver PROJECT_NAME.beatconfig:application

Putting all this information in the beatconfig.py file is necessary to get the channel_layer. This is due to the fact that in Channels 2 if you try to import the PROJECT_NAME.asgi file as was done for Channels 1 you get a conflict between a the reactor imported in Server.py and an asyncioreactor used by Daphne. Daphne notices the conflict and proceeds to uninstall automatically the synchronous beatserver reactor which breaks beatserver, so I had to move the necessary information to get the channel_layers to the beatconfig.py file.... I know that this quite inelegant and you'll probably want to find a better solution before merging my changes, but I thought you'd prefer to at least have some "first working solution" you can improve upon :-).

facucosta committed:
Channels 2 support

These changes (with the ones I propose for server.py) give beatserver Channels 2 support without dropping Channels 1 support. Admittedly the modifications are a bit hacky due to my inexperience and the difference between Channels 1 and Channels 2 APIs, so maybe it would make more sense to fork the project into a Channels-1 version and a Channels-2 version to keep the code clean...

In addition to the changes in these files, the beatserver.py file requires additional lines to work correctly in Channels 2 (no changes needed if the user wants to keep working with Channels 1) which should be documented in the README file. This is an example beatserver.py file that works with Channels 2:

import os
from datetime import timedelta
from channels.layers import get_channel_layer

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "PROJECT_NAME.settings")

BEAT_SCHEDULE = {
    'send-every-5-seconds': {
        'channel_name': 'CHANNEL_NAME',
        'schedule': timedelta(seconds=5),
        'message': {'type': 'type_of_message', 'message': 'message_text'}
    },
}

channel_layers = get_channel_layer()

and the server needs to be initialized with the following command:

beatserver PROJECT_NAME.beatconfig:application

Putting all this information in the beatconfig.py file is necessary to get the channel_layer. This is due to the fact that in Channels 2 if you try to import the PROJECT_NAME.asgi file as was done for Channels 1 you get a conflict between a the reactor imported in Server.py and an asyncioreactor used by Daphne. Daphne notices the conflict and proceeds to uninstall automatically the synchronous beatserver reactor which breaks beatserver, so I had to move the necessary information to get the channel_layers to the beatconfig.py file.... I know that this quite inelegant and you'll probably want to find a better solution before merging my changes, but I thought you'd prefer to at least have some "first working solution" you can improve upon :-).

facucosta committed:
Update server.py

rajasimon commented:
I don't want channel-1 and channel-2 support live in the code base. I like the upcoming release only support Channel 2 with Python 3+ is awesome. Can you make the changes and I will merge.

facucosta commented:
Alright. I will work on it :-).

On 19 March 2018 at 07:33, Raja Simon <notifications@github.com> wrote:

> I don't want channel-1 and channel-2 support live in the code base. I like
> the upcoming release only support Channel 2 with Python 3+ is awesome. Can
> you make the changes and I will merge.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rajasimon/beatserver/pull/6#issuecomment-374167676>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AhLFwTsN9LftZbRdlZDW_FC0pILaeaayks5tf4mSgaJpZM4Sawii>
> .
>



-- 
Facundo Costa

facucosta committed:
Update parser.py

facucosta committed:
Update server.py

facucosta committed:
Update README.md

facucosta committed:
Update README.md

facucosta commented:
Removed Channels 1 support and updated the documentation on how to configure and run it. To be 100% honest I stopped using beatserver for my project (I now run threads in parallel from inside the django project itself since Channels 2 allows to do everything with a single process) so I wasn't able to test this thoroughly myself. It is identical to the previous version removing the Channels 1 "Ifs" parts so there is really no reason whatsoever for it to fail... but I would run a quick test before accepting the pull request if I were you just to be 100% safe that there are no obvious bugs lying around.

rajasimon commented:
I don't know what is the feature of beatserver. Actually I'm not migrated into Channel 2 yet.  Let's open this pull request for awhile. And if people really needs the merge and then I will do it. 

And thanks @facucosta for this merge request. If I do I will give you the full credit bro.

rajasimon commented:
@facucosta I'm trying to implement different approach to this problem. Channel 2.0 by default support background task so I'm trying to implement the same idea with beatserver and There is async sleep support is awesome. Checkout the example code and let me know in the comment. 

It's just the base idea I will clean up the code once fully done.
